### PR TITLE
feat!: `mdc.config.js` support, improve shiki bundle

### LINKED ISSUE #124: Proposal: Support `mdc.config.ts` file
Different from build-based tools like VitePress, one benefit of Nuxt Content/Nuxt MDC is that the rendering could happen on the build/server/client sides isomorphicly for dynamic content. However, with the current approach, the capability of customization is quite limited and not taking full advantage of integrations like `rehype`, `remark` and `shikiji.` 

For example, according to https://content.nuxt.com/get-started/configuration#markdown, to add custom retype/remark plugins, we use a static map/array to define the package name and options. This is limited to only the plugins that have been published to npm, and make it hard to add inline ad-hoc plugins, or high-order plugins. Similar situation applies to Shikiji, where we don't provide options for adding custom transformers.

The limitation coming from the `nuxt.confg.ts` is at built-time, and it's almost impossible to serialize functions and JS closures into the runtime context. Thus, I propose that we introduce a new `mdc.config.ts` file that is intended to be imported by both the client and server sides, similar to `app.config.ts` - so that users could provide customization in a much more interactive way.

For example:

```ts
// mdc.config.ts
import { defineConfig } from '@nuxtjs/mdc/config'
import rehypeDocument from 'rehype-document'
import rehypeFormat from 'rehype-format'
import rehypeRaw from 'rehype-raw'
import rehypeStringify from 'rehype-stringify'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { transformerNotationDiff } from 'shikiji-transformers'

export default defineConfig({
  rehypePlugins: [
    rehypeDocument,
    rehypeForma,
	() => {
       // custom inline plugins
    }
  ],

  remarkPlugins: [
    remarkRehype,
    // ...
  ]

  highlight: async (code, lang) => {
    // custom highlight function
  },

  // or configure as an option with the default highlighter
  highlight: {
    shikiji: {
      transformers: [
        transformerNotationDiff(),
      ]
    }
  }
})
```

And multiple `mdc.config.ts` might be able to be provided via other modules by hooks,

```ts
nuxt.hook('mdc:config-sources', files => {
	files.push('./my-custom-mdc-extension.ts')
})
```

This would allow for integrating features like twoslash #120, provided by a separate module, without everybody paying the dependencies sizes.

### PR DESCRIPTION
**Pull request by antfu**
Resolves #124

## Changes

- Support `mdc.config.ts/js` file, imported via `#mdc-configs`
  - The interface looks like this: https://github.com/nuxt-modules/mdc/blob/feat/mdc-configs/src/config.ts
- With `highlighter: 'custom'`, it allows users to provide a highlighter function (instead of filepath) to have custom highlight. In that case, Shiki won't be included in the bundle at all.
- Now the shiki instance is constructed with [fine-grain bundle](https://shikiji.netlify.app/guide/install#fine-grained-bundle) at build time; This allows MDC to only ships used languages/themes to the bundle, not all of them, making it works more efficient on edge. Accordingly, all languages need to be specified via `highlighter.langs` statically (it's a breaking change). `highlighter.preload` is deprecated.
  - The `.output` dist of the `/playground` comparison:
    - `main` branch: 22MB (4.5MB zipped)
    - this PR: 13MB (3.3Mb zipped)

## Refactor

- Move templates to `/templates` folder.
- Refactored how highlighter is resolved.

## TODOs

- [x] Hooks to add config sources
- [x] Passing code frame meta string to the pipeline
- [x] Test compatibility with downstream Nuxt Content and ecosystem

### CODE REVIEW

**Comment by antfu** [comment]
This PR is ready to review. Also did a test on nuxt.com with: https://github.com/nuxt/nuxt.com/pull/1486

**Comment by farnabaz** [comment]
From what I understand now, we switched to use WASM by default. 

We need to consider a couple of usages:
- MDC module should be able to parse & render contents in browser, with or without API
- MDC module could be used in non-Nuxt context, like Strapi API. We should propose a solution for those environments.
  - Importing `getMdcConfigs` from `#mdc-configs` will fail outside of Nuxt.
  - We need a way to pass `mdcConfig` into the parser function in this case

**Comment by antfu** [comment]
For WASM, Shiki/Shikiji always needs WASM, and the do works in the browser. This PR does not change that.

> MDC in non-Nuxt context

Can you give a more concrete example of what the usages be like? I am happy to expose the necessary options to make them decoupled.

File: src/runtime/highlighter/shiki.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+}: CreateShikiHighlighterOptions): Highlighter {
+  async function _getShiki() {
+    const shiki = await getHighlighterCore({
+      langs,
+      themes,
+      loadWasm: () => import('shiki/wasm')
+    })
+
+    const configs = await getMdcConfigs()
+    for (const config of configs) {
```
Comment (farnabaz) [inline src/runtime/highlighter/shiki.ts:34]:
```suggestion
    for await (const config of configs) {
```

File: src/runtime/highlighter/shiki.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  options: shikiOptions
+}: CreateShikiHighlighterOptions): Highlighter {
+  async function _getShiki() {
+    const shiki = await getHighlighterCore({
+      langs,
+      themes,
+      loadWasm: () => import('shiki/wasm')
+    })
+
+    const configs = await getMdcConfigs()
```
Comment (farnabaz) [inline src/runtime/highlighter/shiki.ts:33]: I think we might also handle missing `getMdcConfigs` and `shikiOptions`

File: src/runtime/parser/index.ts [hunk 3]
```diff
@@ -9,44 +9,91 @@ import { compileHast } from './compiler'
...
+export const parseMarkdown = async (md: string, inlineOptions: MDCParseOptions = {}) => {
   if (!moduleOptions) {
-    // @ts-ignore
     moduleOptions = await import('#mdc-imports' /* @vite-ignore */).catch(() => ({}))
   }
-  const options = defu(opts, {
+  if (!generatedMdcConfigs) {
+    generatedMdcConfigs = await import('#mdc-configs' /* @vite-ignore */)
+      .then(r=>r.getMdcConfigs())
+      .catch()
```
Comment (farnabaz) [inline src/runtime/parser/index.ts:26]:
```suggestion
      .catch(() => ([]))
```

File: src/templates/mdc-highlighter.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      ...Array.from(langs)
+        .map((lang) => typeof lang === 'string'
+          ? `  import('shiki/langs/${lang}.mjs'),`
+          : '  ' + JSON.stringify(lang) + ','),
+      ']',
+      'const themes = [',
+      ...themes.map((theme: string | ThemeRegistration) => typeof theme === 'string'
+        ? `  import('shiki/themes/${theme}.mjs'),`
+        : '  ' + JSON.stringify(theme) + ','),
+      ']',
```
Comment (farnabaz) [inline src/templates/mdc-highlighter.ts:68]: This is great for built time, but I love to have a simple way to load themes and langs dynamically
Currently, I had to this in order to access shiki core and load themes based on highlight request
```ts
let shikiCore = null
  let setupThemes = null

  const shikiHighlighter = createShikiHighlighter({
    options: {},
    getMdcConfigs: () => {
      return [
        {
          shiki: {
            async setup (core) {
              shikiCore = core
              await importThemes(core, setupThemes)
            }
          }
        }
      ]
    }
  })
  const importThemes = async (core, theme) => {
    if (!theme) {
      return
    }
    if (typeof theme === 'string') {
      theme = { default: theme }
    }
    const loadedThemes = core.getLoadedThemes()
    for await (const themeName of Object.values(theme)) {
      if (!loadedThemes.includes(themeName)) {
        await core.loadTheme(import(`shiki/themes/${themeName}.mjs`).then(m => m.default || m))
      }
    }
  }
  const codeHighlighter = async (code, lang, theme, options) => {
    setupThemes = theme
    if (shikiCore) {
      await importThemes(shikiCore, setupThemes)
    }
    return shikiHighlighter(code, lang, theme, options)
  }
```

**Review by farnabaz — APPROVED** [review]
🚀

**Pull request merged**